File: unittest/scripts/auto/js_shell/scripts/binary_data_handling_norecord.js
shell.connect(__mysqluripwd)

let binary_types = ["TINYBLOB", "BLOB", "MEDIUMBLOB", "LONGBLOB", "VARBINARY(20)", "BINARY(30)", "GEOMETRY"]
let custom_values={}

session.runSql("drop schema if exists js_binary_data")
session.runSql("create schema js_binary_data")

function str2ab(str) {
    var buf = new ArrayBuffer(str.length * 2);
    var bufView = new Uint16Array(buf);
    for (var i = 0; i < str.length; i++) {
      bufView[i] = str.charCodeAt(i);
    }
    return buf;
}

function hex2ab(str) {
    var buf = new Uint8Array(str.match(/../g).map(h=>parseInt(h,16)));
    return buf.buffer;
}

// On BINARY(30) the row will be filled with 0's to fulfill the row size, so even the same
// data is stored, a different ArrayBuffer is returned
custom_values["BINARY(30)"] = [str2ab("some\0value"), str2ab("some\0value\0\0\0\0\0"), "0x73006F006D0065000000760061006C007500650000000000000000000000"]
custom_values["GEOMETRY"] = [hex2ab('000000000101000000000000000000F03F000000000000F03F'), hex2ab('000000000101000000000000000000F03F000000000000F03F'), '0x000000000101000000000000000000F03F000000000000F03F']

if (__version_num >= 90000) {
    binary_types.push("VECTOR");
    custom_values["VECTOR"]=[hex2ab("0000803F0000004000004040"), hex2ab("0000803F0000004000004040"), "0x0000803F0000004000004040"];
}


function ab2str(buf) {
    return String.charAt.apply(null, new Uint16Array(buf));
}

function validate(row, binary_type, expected_array_buffer, expected_str) {
    EXPECT_EQ("ArrayBuffer", String(type(row[0])), binary_type)
    EXPECT_EQ(expected_array_buffer, row[0], binary_type)
    // Value as the shell prints it
    print(binary_type, row)
    EXPECT_STDOUT_CONTAINS(expected_str, binary_type)
    // Value as printed by python
    print(row[0])
    EXPECT_STDOUT_CONTAINS(expected_str, binary_type)
}

function test_runSql(asession, binary_type, array_buffer, expected_array_buffer, expected_str) {
    asession.runSql("insert into js_binary_data.sample values (_binary ?)", [array_buffer])

    res = asession.runSql("select * from js_binary_data.sample")
    row = res.fetchOne()
    validate(row, binary_type, expected_array_buffer, expected_str)
    session.runSql("delete from js_binary_data.sample")
}


function test_classic(binary_type, array_buffer, expected_array_buffer, expected_str) {
    asession = mysql.getSession(__mysqluripwd)
    test_runSql(asession, binary_type, array_buffer, expected_array_buffer, expected_str)
    asession.runSql("delete from js_binary_data.sample")
    asession.close()
}

function test_table_crud(asession, binary_type, array_buffer, expected_array_buffer, expected_str) {
    schema = asession.getSchema('js_binary_data')
    table = schema.sample
    table.insert().values(array_buffer).execute()
    row = table.select().execute().fetchOne()
    validate(row, binary_type, expected_array_buffer, expected_str)
}

function test_x(binary_type, array_buffer, expected_array_buffer, expected_str) {
    asession = mysqlx.getSession(__uripwd)
    test_runSql(asession, binary_type, array_buffer, expected_array_buffer, expected_str)
    asession.runSql("delete from js_binary_data.sample")
    test_table_crud(asession, binary_type, array_buffer, expected_array_buffer, expected_str)
    asession.runSql("delete from js_binary_data.sample")
}

for (index in binary_types) {
    binary_type = binary_types[index]
    // An array buffer will be created to hold the data to be stored on the database
    // By default it is expected that once the data is retrieved an identical array buffer
    // is returned except for BINARY(30), see below.
    let array_buffer = str2ab("some\0value")
    let expected_array_buffer = array_buffer
    expected_str = "0x73006F006D0065000000760061006C0075006500"

    if (binary_type in custom_values) {
       array_buffer = custom_values[binary_type][0]
       expected_array_buffer = custom_values[binary_type][1]
       expected_str = custom_values[binary_type][2]
    }

    session.runSql(`create table js_binary_data.sample(data ${binary_type})`)
    // Test data insertion and retrieval using ClassicSession.runSql
    test_classic(binary_type, array_buffer, expected_array_buffer, expected_str)
    if (binary_type != "VECTOR") // NOTE: not supported atm
        test_x(binary_type, array_buffer, expected_array_buffer, expected_str)
    session.runSql("drop table js_binary_data.sample")
}

session.runSql("drop schema js_binary_data")
shell.disconnect()

File: unittest/scripts/auto/py_shell/scripts/binary_data_handling_norecord.py
shell.connect(__mysqluripwd)

binary_types = ["TINYBLOB", "BLOB", "MEDIUMBLOB", "LONGBLOB", "VARBINARY(20)", "BINARY(20)"]
custom_values={"BINARY(20)": (b'some\x00value\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', "0x736F6D650076616C756500000000000000000000", "b'some\\x00value\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00'")}

session.run_sql("drop schema if exists py_binary_data")
session.run_sql("create schema py_binary_data")


def validate(row, binary_type, expected_binary_value, expected_str_shell, expected_str_python):
    EXPECT_EQ("<class 'bytes'>", str(type(row[0])), binary_type)
    EXPECT_EQ(expected_binary_value, row[0], binary_type)
    # Value as the shell prints it
    print(row)
    EXPECT_STDOUT_CONTAINS(expected_str_shell, binary_type)
    # Value as printed by python
    print(row[0])
    EXPECT_STDOUT_CONTAINS(expected_str_python, binary_type)

def test_run_sql(asession, binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python):
    # Test data insertion and retrieval using ClassicSession.run_sql
    asession.run_sql("insert into py_binary_data.sample values (?)", [b'some\0value'])
    res = asession.run_sql("select * from py_binary_data.sample")
    row = res.fetch_one()
    validate(row, binary_type, expected_binary_value, expected_str_shell, expected_str_python)
    session.run_sql("delete from py_binary_data.sample")


def test_classic(binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python):
    asession = mysql.get_session(__mysqluripwd)
    test_run_sql(asession, binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python)
    asession.run_sql("delete from py_binary_data.sample")
    asession.close()

def test_table_crud(asession, binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python):
    schema = asession.get_schema('py_binary_data')
    table = schema.sample
    table.insert().values(binary_value).execute()
    row = table.select().execute().fetch_one()
    validate(row, binary_type, expected_binary_value, expected_str_shell, expected_str_python)


def test_x(binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python):
    asession = mysqlx.get_session(__uripwd)
    test_run_sql(asession, binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python)
    asession.run_sql("delete from py_binary_data.sample")
    test_table_crud(asession, binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python)
    asession.run_sql("delete from py_binary_data.sample")
    asession.close()

for binary_type in binary_types:
    # An binary value will be created to hold the data to be stored on the database
    # By default it is expected that once the data is retrieved an identical binary value
    # is returned except for BINARY(30), see below.
    binary_value=b'some\x00value'
    expected_binary_value = binary_value
    expected_str_shell = "0x736F6D650076616C7565"
    expected_str_python = "b'some\\x00value'"
    # On BINARY(20) the row will be filled with 0's to fulfill the row size, so even the same
    # data is stored, a different binary value is returned
    if binary_type in custom_values:
       expected_binary_value, expected_str_shell, expected_str_python = custom_values[binary_type]
    session.run_sql(f"create table py_binary_data.sample(data {binary_type})")
    # Test data insertion and retrieval using ClassicSession.run_sql
    test_classic(binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python)
    test_x(binary_type, binary_value, expected_binary_value, expected_str_shell, expected_str_python)
    session.run_sql("drop table py_binary_data.sample")

session.run_sql("drop schema py_binary_data")
shell.disconnect()

#Shell_scripted/Auto_script_py.run_and_check/binary_data_handling